### PR TITLE
Ensure that the URI of the response matches the request

### PR DESCRIPTION
**Pull request by picsel2**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 

## The problem

The `textDocument/publishDiagnostics` notification modifies the file URI if its path contains a symbolic link.

How to reproduce:

1. Create a workspace where the project root is a symbolic link to another directory containing a markdown file.
2. Ensure that the file URIs sent by the editor are not resolved.
3. The `textDocument/publishDiagnostics` notification contains the resolved file path.

## Expectation

File URIs should be preserved and not resolved.

## Description of changes

I used the VFile.data property to store the original document URI. I don't like that solution much but it is the best I could find without doing big changes.

I also tried hard to add a test but were not able to accomplish that with reasonable effort.

### CODE REVIEW

**Comment by wooorm** [comment]
- Why is what you describe as “problem”, a problem?
- Why should symlinks (not) be resolved?
- I don’t think any explicit call here resolves symlinks, so, where is it coming from?
- Files already have history: `history[0]` refers to the original file
- There should be a test. What did you try?

**Comment by picsel2** [comment]
> * Why is what you describe as “problem”, a problem?

I tried to configure the emacs lsp-mode package to use remark-language-server for markdown files. I ran into this problem because one of these files is opened via a symlinked directory.

My concrete situation might be reproduced as follows:
```sh
mkdir ~/git/emacs-config
echo "Some content that produces diagnostics messages" > ~/git/emacs-config/README.md
ln -s ~/git/emacs-config ~/.config/emacs
emacs ~/.config/emacs/README.md # open the file using the symlinked directory
```

The result is that lsp-mode does not show any of these diagnostics.

[Here](https://gist.github.com/picsel2/3c5584ea84d6ec27aa8707177ba9afe0#file-gistfile1-txt) is the logged communication between lsp-mode and remark-language-server.

It shows that lsp-mode refers to the unresolved path (which is what I would also expect) in [its URIs](https://gist.github.com/picsel2/3c5584ea84d6ec27aa8707177ba9afe0#file-gistfile1-txt-L272). On the other hand the remark-language-server sends back [resolved URIs](https://gist.github.com/picsel2/3c5584ea84d6ec27aa8707177ba9afe0#file-gistfile1-txt-L520) as part of the `textDocument/publishDiagnostics` notification.

I think it is reasonable that the lsp-mode then "loses" the association with the opened file.

> * Why should symlinks (not) be resolved?

Resolving the symlinks leads to confusion. The association with the file gets lost because resolving the symlink changes the resource identifier. So from the view of the editor the URIs `file:///home/sebastian/.config/emacs/README.md` and `file:///home/sebastian/git/emacs-config/README.md` are two different files.

Unfortunately I could not find anything in the LSP spec on how symbolic links should be handled.

> * I don’t think any explicit call here resolves symlinks, so, where is it coming from?

[VFile.path](https://www.npmjs.com/package/vfile#vfilepath) gets called [in the checkDocuments function](https://github.com/unifiedjs/unified-language-server/blob/dba7db68db8186a969f7affc5bf5a2c4bf7793bc/lib/index.js#L311). The documentation of VFile states explicitly that is uses `url.fileURLToPath`, which does the actual resolving.

> * Files already have history: `history[0]` refers to the original file

As far as I understand these are already resolved paths like the VFile.path

> * There should be a test. What did you try?

I tried to find out how to write an isolated test. But the api only allows to start a full-fledged lsp server using a subprocess (as it is done in the already existing tests). So to me it seems that the only way to test this is to "reverse engineer" the already existing test patterns and try to trigger a `textDocument/publishDiagnostics` notification on a prepared file. This goes way beyond my understanding of the protocol and my available time.

**Comment by remcohaszing** [comment]
I see the problem, but I’m not too fond of the solution. I can’t tell how vfile deals with symlinks, so if this is an issue you ran into, I’m inclined to believe vfile resolves them. This is certainly worth investigating!

Could you add a test for this?

**Comment by wooorm** [comment]
> It shows that lsp-mode refers to the unresolved path in its URIs. On the other hand the remark-language-server sends back resolved URIs as part of the textDocument/publishDiagnostics notification.

While your problem currently presents itself around symlinks, it might not be a problem unique to symlinks (e.g., there might be other values that are sent which are sent back differently, due to escapes/encoding of characters, slashes, case-sensitive/-insensitive file systems, &c.). Perhaps it’s better to approach (and solve) this problem more generally?

> The documentation of VFile states explicitly that is uses `url.fileURLToPath`, which does the actual resolving.

Resolving symlinks requires access to the file system. `vfile` (and `node:url`, which contains `fileURLToPath`) **does not access the file system**.
This project **does** access the file system and resolve symlinks. Because it must try and find that file on the file system. And it has to configure it. But that’s fine — it’s supposed to do that.

> As far as I understand these are already resolved paths like the `VFile.path`

Yep. But that has no effect on symlinks.

> So to me it seems that the only way to test this is to "reverse engineer" the already existing test patterns and try to trigger a textDocument/publishDiagnostics notification on a prepared file. This goes way beyond my understanding of the protocol and my available time.

There is no reverse engineering needed.
You can write a symlink to the file system.
Or you can try spaces and encoding in tests. Or something similar.
Tests are required for this to land though.

---

For a solution, I think we can solve this quite like what you’re currently doing it. But I dislike the name `file.data.uri`. And commenting about symlinks only.
What about calling it `file.data.lspInputUri` or so? To signal that its purpose is to pass a certain string from that protocol through the engine and back?

**Comment by remcohaszing** [comment]
> While your problem currently presents itself around symlinks, it might not be a problem unique to symlinks (e.g., there might be other values that are sent which are sent back differently, due to escapes/encoding of characters, slashes, case-sensitive/-insensitive file systems, &c.). Perhaps it’s better to approach (and solve) this problem more generally?

I didn’t think of this of being the same problem, but you’re right, maybe it is. E.g. the Windows drive letter casing is also causing issues.

> For a solution, I think we can solve this quite like what you’re currently doing it. But I dislike the name `file.data.uri`. And commenting about symlinks only.  
> What about calling it `file.data.lspInputUri` or so? To signal that its purpose is to pass a certain string from that protocol through the engine and back?

My concern is about a potentially conflict in data fields with a potenrial third party plugin. I agree a more specific name such as `lspInputUri` (or `lspDocumentUri`?) resolves this issue. Perhaps a symbol could be used.

**Comment by picsel2** [comment]
I think I've been mislead by the documentation of `url.fileURLToPath`. While it says "fully-resolved" path, is does not seem to resolve symlinks.

The snippet
```js
import url from 'node:url';
import { VFile } from 'vfile';
console.log(url.fileURLToPath(new URL("file:///home/sebastian/.config/emacs")))
console.log(new VFile(new URL("file:///home/sebastian/.config/emacs")))
```
outputs unresolved symlinks
```
/home/sebastian/.config/emacs
VFile {
  data: {},
  messages: [],
  history: [ '/home/sebastian/.config/emacs' ],
  cwd: '/home/sebastian/temp/nodetest'
}
```

With some printf-debugging I found out, that the actual "resolving" comes from passing only the file's basename to  `url.pathToFileURL` which then gets appended to the current working directory.
It seems that engine returns the VFiles with only their basename and other cwds:
```
files passed to engine:
[
  VFile {
    data: {},
    messages: [],
    history: [ '/home/sebastian/.config/emacs/README.md' ],
    cwd: '/home/sebastian/git/emacs-config',
    value: <...>
  }
]
result returned from engine:
[
  VFile {
    data: { unifiedEngineGiven: true },
    messages: [
      [VFileMessage],
      [VFileMessage],
      [VFileMessage],
      [VFileMessage],
      [VFileMessage],
      [VFileMessage],
      [VFileMessage],
      [VFileMessage],
      [VFileMessage],
      [VFileMessage]
    ],
    history: [ 'README.md' ],
    cwd: '/home/sebastian/.config/emacs',
    value: <...>
  }
]
```

**Comment by wooorm** [comment]
In response to remco:

> E.g. the Windows drive letter casing is also causing issues.

did you practically notice this before, or is this a guess? If so, how?

> My concern is about a potentially conflict in data fields with a potenrial third party plugin.

There’s some precedence: https://github.com/unifiedjs/unified-engine/blob/aa1f1d96ad3e029a920bad75e7f224d1946d8083/lib/file-pipeline/stdout.js#L22. You’re right that it could be problematic at some point, but it’s also useful and exactly what that space is for.

> I agree a more specific name such as lspInputUri (or lspDocumentUri?)

`lspDocumentUri` is great, yeah! I like it more than `lspInputUri`

> Perhaps a symbol could be used.

I prefer sticking with plain JSON! I don’t see it practically becoming a problem!

---

In response to Sebastian:

> While it says "fully-resolved" path, is does not seem to resolve symlinks.

Resolved has several interpretations I guess. In this case it’s more about `./path/../file.html` -> `./file.html` or so.
Also quite luckily that Node.js is not Java: https://news.ycombinator.com/item?id=21765788 ;)

So, Sebastian, can you switch to use `lspDocumentUri`? Add a test that creates a symlink / spaces / casing / unicode or so?

**Comment by remcohaszing** [comment]
> > E.g. the Windows drive letter casing is also causing issues.
> 
> did you practically notice this before, or is this a guess? If so, how?

Yes, I mentioned it in #29

File: lib/index.js [hunk 1]
```diff
@@ -307,8 +314,9 @@ export function configureUnifiedLanguageServer(
     const files = await processDocuments(textDocuments)
 
     for (const file of files) {
-      // VFile uses a file path, but LSP expects a file URL as a string.
-      const uri = String(pathToFileURL(file.path))
+      // VFile.path for URI restoration is not suitable here because it
+      // resolves symlinks
+      const uri = String(file.data.uri)
```
Comment (remcohaszing) [inline lib/index.js:319]: I have also been fiddling with this, because this turns out to be very closely related to #29.

I think this specific issue can be resolved using the following change:

```suggestion
      // unified-engine returns a relative path, but LSP expects an absolute file URL as a string.
      const uri = String(pathToFileURL(path.resolve(file.cwd, file.path)))
```

This means the custom data can be removed.

Comment (remcohaszing) [reply lib/index.js:319]: I changed my mind about this. Due to various internal transforms (url ↔ path, `/` ↔ path separator, case insensitive file systems), I now think it makes perfect sense to keep the original document uri around to use for reporting back.

**Comment by wooorm** [comment]
@picsel2 Can you explain why you closed this? Is this no longer an issue for you?

**Comment by picsel2** [comment]
@wooorm I simply lost the interest. I don't know if this bug is still there or not, because I didn't not check again.

**Comment by wooorm** [comment]
I understand, thanks. I’ve opened GH-38 instead.
